File: logon_digital_hangzhou_dianzi_university/logon.php
<?php
header("Content-Type: text/html; charset=utf-8");

$username = "xxx";
$password = "xxx";

$cookie=dirname(__FILE__).'/cookie.txt';

#先访问一次获取参数lt
$ch = curl_init(); 
curl_setopt($ch, CURLOPT_URL, 'http://cas.hdu.edu.cn/cas/login'); 
curl_setopt($ch, CURLOPT_HEADER, 0);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
$res=curl_exec($ch);
curl_close($ch);
$preg = '|<input type="hidden" name="lt" value=[\"](.*?)[\"] />|U';
preg_match_all($preg, $res, $arr); 
$lt=$arr[1][0];

#登录，模拟他的post请求
$post_data=array(
    'encodedService'=>'http%3a%2f%2fi.hdu.edu.cn%2fdcp%2findex.jsp',
    'service'=>'http://i.hdu.edu.cn/dcp/index.jsp',
    'serviceName'=>'null',
    'loginErrCnt'=>'0',
    'username'=>$username,
    'password'=>md5($password),
    'lt'=>$lt
);
$curl = curl_init();
curl_setopt($curl, CURLOPT_URL, "http://cas.hdu.edu.cn/cas/login");
curl_setopt($curl, CURLOPT_HEADER, 0);
curl_setopt($curl, CURLOPT_RETURNTRANSFER, 1);
curl_setopt ($curl,CURLOPT_REFERER,'http://cas.hdu.edu.cn/cas/login');
curl_setopt($curl, CURLOPT_USERAGENT, 'Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/45.0.2454.101 Safari/537.36'); 
curl_setopt($curl, CURLOPT_POST, 1);//post方式提交 
curl_setopt($curl, CURLOPT_COOKIEJAR, $cookie);
curl_setopt($curl, CURLOPT_POSTFIELDS, http_build_query($post_data));//要提交的信息 
curl_setopt($curl, CURLOPT_FOLLOWLOCATION, 3);       

#这里成功会返回一个window.location.href跳转
$res=curl_exec($curl);//执行cURL 
curl_close($curl);

#匹配跳转，也可以带cookie继续访问进去
preg_match_all('/window.location.href=[\"](.*)[\"]/i', $res, $results);
if(is_null($results[1][0])){
    echo "用户名密码错误";
}else{
    echo "已登录，可继续带cookie访问 ".$results[1][0];
}

File: logon_digital_hangzhou_dianzi_university/logon.py
import requests
import re
import hashlib

#先访问一次登录网站得到lt（lt后面必须，且一次性使用）
def getHduCookie():
    resp=requests.get('http://cas.hdu.edu.cn/cas/login')
    m = re.search(r'name=\"lt\" value=(.*?) />', resp.text)
    lt=m.group()[17:-4]
    return lt

#模拟登陆用户名为学号，密码为md5加密后的密码，返回跳转链接
def simLogin(lt):
    password=hashlib.md5(psd.encode('utf-8')).hexdigest()
    params={
        'encodedService':'http%3a%2f%2fi.hdu.edu.cn%2fdcp%2findex.jsp',
        'service':'http://i.hdu.edu.cn/dcp/index.jsp',
        'username':xh,
        'password':password,
        'lt':lt
    }
    resp=requests.post('http://cas.hdu.edu.cn/cas/login?service=http://jxgl.hdu.edu.cn/index.aspx',params=params)
    m=re.search(r'href="(.*?)"',resp.text)
    return m.group()[6:-1]

#去临时链接获取一次cookie并保存请求
def jxglPage(url):
    req=requests.Session()
    resp0=req.get(url)
    req.headers={
        'User-Agent':'Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/56.0.2924.87 Safari/537.36',
        'Referer':'http://jxgl.hdu.edu.cn/xf_xsqxxxk.aspx?xh='+xh+'&xm=%25%5cB8%25%5cCA%25%5cBC%25%5cD2%25%5cB3%25%5cC7&gnmkdm=N121113',
        'Accept-Encoding': 'gzip, deflate',
        'Content-Type':'application/x-www-form-urlencoded',
    }
    return req

#这里的例子是获取选课列表
def classList(req):
    data=''#这里是post的一大串字符，可从浏览器获取
    url='http://jxgl.hdu.edu.cn/xf_xsqxxxk.aspx?xh='+xh+'&xm=%25%5cB8%25%5cCA%25%5cBC%25%5cD2%25%5cB3%25%5cC7&gnmkdm=N121113'
    resp=req.post(url,data=data)
    resp.encoding='gbk'
    print(resp.text)

#主程序，设置学号密码并运行
if __name__=='__main__':
    xh='xxx'
    psd='xxx'
    lt=getHduCookie()
    tmpurl=simLogin(lt)
    if tmpurl[0:4]=="http":
        print("登录成功，可跳转网址"+tmpurl)
    else:
        print("登录失败")
    #req=jxglPage(tmpurl)
    #classList(req)